### PR TITLE
refactor(tcp): reducing branching in `Transport::create_socket`

### PR DESCRIPTION
**Pull request by jxs**
and add test for listening on ipv4 and ipv6 separately.

## Description

Following https://github.com/libp2p/rust-libp2p/pull/4289#discussion_r1285812762, hereby is the PR to also improve the `create_socket`  using [`for_addr`](https://docs.rs/socket2/latest/socket2/struct.Domain.html#method.for_address). We also add a test for listening on IPv4 and IPv6 separately.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/tcp/src/lib.rs [hunk 1]
```diff
@@ -1329,4 +1328,35 @@ mod tests {
...
+                test::<async_io::Tcp>(3001);
+            })
+        }
+        let rt = ::tokio::runtime::Builder::new_current_thread()
+            .enable_io()
+            .build()
+            .unwrap();
+        rt.block_on(async {
+            test::<async_io::Tcp>(4001);
+        });
```
Comment (mxinden) [inline transports/tcp/src/lib.rs]: Wouldn't the tokio execution also be feature flagged? In other words, does `cd transports/tcp && cargo test --feature async-io` succeed?

**Review by mxinden — COMMENTED** [review]
Thank you for the follow-up.

File: transports/tcp/src/lib.rs (hunk 1 above)
Comment (jxs) [reply transports/tcp/src/lib.rs]: no, right. Thanks for the correction Max, updated.

File: transports/tcp/src/lib.rs [hunk 2]
```diff
@@ -1329,4 +1328,38 @@ mod tests {
...
+            tcp.listen_on(
+                ListenerId::next(),
+                format!("/ip6/::/tcp/{port}").parse().unwrap(),
+            )
+            .unwrap();
+        }
+        #[cfg(feature = "async-io")]
+        {
+            async_std::task::block_on(async {
+                test::<async_io::Tcp>(3001);
```
Comment (mxinden) [inline transports/tcp/src/lib.rs]:
```suggestion
                test::<async_io::Tcp>(0);
```

File: transports/tcp/src/lib.rs [hunk 3]
```diff
@@ -1329,4 +1328,38 @@ mod tests {
...
+            })
+        }
+        #[cfg(feature = "tokio")]
+        {
+            let rt = ::tokio::runtime::Builder::new_current_thread()
+                .enable_io()
+                .build()
+                .unwrap();
+            rt.block_on(async {
+                test::<async_io::Tcp>(4001);
```
Comment (mxinden) [inline transports/tcp/src/lib.rs]:
```suggestion
                test::<async_io::Tcp>(0);
```

**Review by mxinden — COMMENTED** [review]
I suggest having the operating system choose any free port. In case someone has e.g. a go-ipfs node running on `4001` their unit tests would still succeed.

File: transports/tcp/src/lib.rs (hunk 3 above)
Comment (mxinden) [inline transports/tcp/src/lib.rs]:
```suggestion
                test::<tokio::Tcp>(4001);
```

**Comment by jxs** [comment]
> I suggest having the operating system choose any free port. In case someone has e.g. a go-ipfs node running on `4001` their unit tests would still succeed.

but that defeats the purpose of the test no? I.e. being certain that ipv4 and ipv6 run in the same port. I can try something else nonetheless.

**Comment by mxinden** [comment]
Ah, yes, sorry. My bad. How about a more random port than `4001`, or even better first assign a random port for v4 by the OS and then use the same port for v6?

**Comment by thomaseizinger** [comment]
You can use a TcpListener to first bind a random port, retrieve the port, drop the listener and then bind to it again :)

**Comment by jxs** [comment]
> You can use a TcpListener to first bind a random port, retrieve the port, drop the listener and then bind to it again :)

yup exactly :grin:. Meanwhile also updated the Quic test, if wanted I can move it instead do a new PR aha

**Review by mxinden — APPROVED** [review]
Neat. Thank you. Fine to merge in one pull request.

**Comment by thomaseizinger** [comment]
@mxinden Why is the `cargo-deny` job suddenly required? It only runs on changes to `Cargo.toml` files which is not the case in all PRs. I can't see it here: https://github.com/libp2p/github-mgmt/blame/master/github/libp2p.yml#L7074-L7126

**Comment by mxinden** [comment]
@mergify refresh

**Comment by mxinden** [comment]
> @mxinden Why is the `cargo-deny` job suddenly required? It only runs on changes to `Cargo.toml` files which is not the case in all PRs. I can't see it here: https://github.com/libp2p/github-mgmt/blame/master/github/libp2p.yml#L7074-L7126

My bad. I changed it through the UI without thinking it all the way through. Thanks for catching. Sorry for the debugging time wasted on this.

What do you think of running it on all changes, simply succeeding right away on non-Cargo.toml changes @thomaseizinger? I would like to get to a place where one does not have to manually check for non-required failures.

**Comment by thomaseizinger** [comment]
> What do you think of running it on all changes, simply succeeding right away on non-Cargo.toml changes @thomaseizinger? I would like to get to a place where one does not have to manually check for non-required failures.

I think we can simply add it to the CI workflow. I don't see why it shouldn't run on all code changes. It is slightly wasteful in terms of resource usage but I don't think it weighs in very heavily for the gained simplicity in the setup.

**Comment by mxinden** [comment]
Sounds good. Let me know what you think of https://github.com/libp2p/rust-libp2p/pull/4352.

**Pull request merged**